tools/ttysnoop.py

```diff
@@ -62,33 +69,75 @@ def usage():
 bpf_text = """
 #include <uapi/linux/ptrace.h>
 #include <linux/fs.h>
+#include <linux/uio.h>
 
-#define BUFSIZE 256
+#define BUFSIZE USER_DATASIZE
 struct data_t {
     int count;
     char buf[BUFSIZE];
 };
 
+BPF_ARRAY(data_map, struct data_t, 1);
 BPF_PERF_OUTPUT(events);
 
+static int do_tty_write(void *ctx, const char __user *buf, size_t count)
+{
+    int zero = 0, i;
+    struct data_t *data;
+
+    data = data_map.lookup(&zero);
+    if (!data)
+        return 0;
+
+    #pragma unroll
+    for (i = 0; i < USER_DATACOUNT; i++) {
+        // bpf_probe_read_user() can only use a fixed size, so truncate to count
+        // in user space:
+        if (bpf_probe_read_user(&data->buf, BUFSIZE, (void *)buf))
+            return 0;
+        if (count > BUFSIZE)
+            data->count = BUFSIZE;
+        else
+            data->count = count;
+        events.perf_submit(ctx, data, sizeof(*data));
+        if (count < BUFSIZE)
+            return 0;
+        count -= BUFSIZE;
+        buf += BUFSIZE;
+    }
+
+    return 0;
+};
+
+/**
+ * commit 9bb48c82aced (v5.11-rc4) tty: implement write_iter
+ * changed arguments of tty_write function
+ */
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 11, 0)
```

Comment:
I forgot to mention last time. bcc tools typically do not use this LINUX_VERSION_CODE comparison approach. The reason is due to backported kernels whose kernel version say is less then 5.11, but may actually have this feature supported. You can search various "supported" clause and may find some examples. Is it possible we could find
some variable/macro/structure we could use to identify this tty_write change?

Also, for ubuntu 16.04 failure issue, I think it is OKAY to only support the new feature with new tty_write signature so the tool can continue to work with old kernels.

Reply:
> I forgot to mention last time. bcc tools typically do not use this LINUX_VERSION_CODE comparison approach. The reason is due to backported kernels whose kernel version say is less then 5.11, but may actually have this feature supported. You can search various "supported" clause and may find some examples. Is it possible we could find
> some variable/macro/structure we could use to identify this tty_write change?

I'll check but I'm not sure

> 
> Also, for ubuntu 16.04 failure issue, I think it is OKAY to only support the new feature with new tty_write signature so the tool can continue to work with old kernels.

hum,  you mean the tool WILL NOT continue to work with older kernels, right? if we support only new tty_write signature

Reply:
No, for the program with *new* function signature, only newer kernel is supported, you are free to use the newer features (like map value for bpf_perf_event_output). The *old* function signature should still be supported, e.g., with the original bpf code. This way, both old kernels and new kernels are supported by the tool.

Reply:
> I'll check but I'm not sure.
Thanks!